### PR TITLE
Fix CriticalSection lock for nrf51 targets.

### PR DESCRIPTION
**Pull request by pan-**
The previous implementation of CriticalSectionLock for nrf51 was only
relying to softdevice primitives sd_nvic_critical_region_enter and
sd_nvic_critical_region_exit .
If the soft device is not enabled, these functions have no effect. This
means that CriticalSectionLock was not locking anything if the softdevice
was not enabled.

This fix allow the usage of CriticalSectionLock, even if the softdevice is
not enabled. If the softdevice is not enabled, all irq are disabled until
the destruction of the lock.

This PR should be merged with: https://github.com/ARMmbed/mbed-hal-nrf51822-mcu/pull/48

### CODE REVIEW

**Comment by pan-** [comment]
@rgrover @0xc0170 could you review this please ?

**Comment by rgrover** [comment]
+1

**Comment by emidttun** [comment]
Looks correct to me, but out of curiosity: Why isn't there an IRQ enable in the destructor? (It was like that before the change.)

**Comment by 0xc0170** [comment]
> Looks correct to me, but out of curiosity: Why isn't there an IRQ enable in the destructor? (It was like that before the change.)

Which line are you referencing ? It restores primask, as it was before.

File: core-util/CriticalSectionLock.h [hunk 1]
```diff
@@ -47,11 +48,29 @@ namespace util {
...
+
 class CriticalSectionLock {
 public:
     CriticalSectionLock() {
 #ifdef TARGET_NORDIC
-        sd_nvic_critical_region_enter(&_state);
+        // get the state of exceptions for the CPU
+        _PRIMASK_state = __get_PRIMASK();
+
+        // if exceptions are not enabled, there is nothing more to do
```
Comment (0xc0170) [inline core-util/CriticalSectionLock.h]: Please fix the formatting in the changes:

an example:
```
        if (_PRIMASK_state == 1) {
            _use_softdevice_routine = false;
        } else {
```

Comment (pan-) [inline core-util/CriticalSectionLock.h]: I've done it.

**Comment by emidttun** [comment]
@0xc0170  So the __disable_irq() only writes to PRIMASK? Wasn't aware of it; then no further comments.

**Comment by 0xc0170** [comment]
LGTM, however it is lot of code for for critical section :-)

cc @bogdanm

**Comment by bogdanm** [comment]
That does look like a lot of code, thus quite a few opportunities for race conditions. I'm going to trust @rgrover on this one, since I'm not that familiar with nRF51 targets. So, +1.

**Pull request merged**